### PR TITLE
Fault tolerance for event bus subscription

### PR DESCRIPTION
**Pull request by tsegismont**
This PR supersedes #1735 

It includes the original commit from @rikardscm as well my updates and a test which cluster manager implementations should run. I've tried it with Hazelcast and Infinispan cluster managers successfully.

Note that with this change, it won't be possible connect older versions to the cluster, because the data format of the the subs multimap is not compatible. As a workaround, one can use the TCP bridge.

### CODE REVIEW

**Comment by tsegismont** [comment]
@vietj can you have a look? When we're fine with the changes, I'll update the other cluster managers as well.

File: src/main/java/io/vertx/core/spi/cluster/AsyncMultiMap.java [hunk 1]
```diff
@@ -63,4 +65,6 @@
    * @param completionHandler This will be called when the remove is complete
    */
   void removeAllForValue(V v, Handler<AsyncResult<Void>> completionHandler);
+
+  void removeAll(Predicate<V> p, Handler<AsyncResult<Void>> completionHandler);
```
Comment (vietj) [inline src/main/java/io/vertx/core/spi/cluster/AsyncMultiMap.java]: should be renamed `removalAllMatching`

File: src/main/java/io/vertx/core/eventbus/impl/clustered/ClusterNodeInfo.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ */
+
+package io.vertx.core.eventbus.impl.clustered;
+
+import java.io.Serializable;
+
+/**
+ * @author Rikard Björklind
+ */
+public class ClusterNodeInfo implements Serializable {
```
Comment (vietj) [inline src/main/java/io/vertx/core/eventbus/impl/clustered/ClusterNodeInfo.java]: can we have ClusterNodeInfo extends or embed ServerID ? that would avoid to recreate a ServerId from ClusterNodeInfo later

**Comment by tsegismont** [comment]
Done

2017-03-01 8:48 GMT+01:00 Julien Viet <notifications@github.com>:

**Comment by tsegismont** [comment]
Done

2017-03-01 8:50 GMT+01:00 Julien Viet <notifications@github.com>:

**Comment by vietj** [comment]
👍 when shall I merge this ?

**Comment by tsegismont** [comment]
I just sent the last follow up PRs for the Ignite and JGroups CM.

This one can be merged when all downstream PRs have been reviewed (HZ,
ISPN, ZK, Ignite, JGroups).

2017-03-01 13:30 GMT+01:00 Julien Viet <notifications@github.com>:

**Comment by AlexThurston** [comment]
Any chances of this being ported to Vert.x 2?

**Comment by vietj** [comment]
@AlexThurston no, we don't maintain anymore the 2.x series

**Pull request merged**